src/main/java/org/apache/commons/lang3/reflect/TypeUtils.java

```diff
@@ -1822,14 +1841,17 @@ private static void appendRecursiveTypes(final StringBuilder buf, final int[] re
             appendAllTo(buf.append('<'), ", ", argumentsFiltered).append('>');
         }
     }
-
-    private static int[] findRecursiveTypes(final ParameterizedType p) {
-        final Type[] filteredArgumentTypes = Arrays.copyOf(p.getActualTypeArguments(), p.getActualTypeArguments().length);
-        int[] indexesToRemove = {};
-        for (int i = 0; i < filteredArgumentTypes.length; i++) {
+    @SuppressWarnings("assignment.type.incompatible")/*
+    #3 Array.copy(array, n) returns an array of length n 
+    #4 i -> @IndexFor("p.getActualTypeArguments()") inside the loop
+    */
+    private static int @SameLen("#1.getActualTypeArguments()") [] findRecursiveTypes(final ParameterizedType p) {
+        final Type @SameLen("p.getActualTypeArguments()") [] filteredArgumentTypes = Arrays.copyOf(p.getActualTypeArguments(), p.getActualTypeArguments().length); // #3
+        int @SameLen("p.getActualTypeArguments()") [] indexesToRemove = {};
```

Comment:
This annotation doesn't seem correct, since the size of `indexesToRemove` can change?